---
layout: default
---

<h1>Tutorials</h1>

<p>Please refer to the following vignettes to help you get started with using <code>scde</code>:</p>

<ul>
  <li><a href="diffexp.html">Single cell differential expression analysis</a></li>
  <li><a href="pagoda.html">Pathway and gene set overdispersion analysis</a></li>
  <li><a href="experimental.html">Experimental use of <code>pagoda</code></a></li>
  <li><a href="genesets.html">Preparing custom gene sets for <code>pagoda</code></a></li>
</ul>

<h1>Additional Resources</h1>

<p>Additional tutorials featuring <code>scde</code> and <code>pagoda</code></p>

<ul>
  <li><a href="http://hms-dbmi.github.io/scw/">Single-Cell Workshop 2015</a></li>
  <li><a href="http://pklab.med.harvard.edu/scw2014/">Single-Cell Workshop 2014</a></li>
</ul>

<p>Browse <code>pagoda</code> apps for public datasets</p>

<ul>
  <li><a href="http://pklab.med.harvard.edu/scde/pagoda.links.html">All datasets analyzed in  <a href="http://www.nature.com/nmeth/journal/vaop/ncurrent/full/nmeth.3734.html">"Characterizing transcriptional heterogeneity through pathway and gene set overdispersion analysis" (Fan J, Salathia N, Liu R, Kaeser G, Yung Y, Herman J, Kaper F, Fan JB, Zhang K, Chun J, and Kharchenko PV, Nature Methods, doi:10.1038/nmeth.3734)</a>
  </li>
</ul>
