---
title: "Pagoda tutorial"
author: "peterk"
date: "3/16/2015"
output: html_document
---

The following tutorial illustrates how to use PAGODA routines (implemented in the scde package) using T-cell data from [_Pollen et al._](www.ncbi.nlm.nih.gov/pubmed/25086649).

```{r,include=FALSE}
.libPaths(c("/home/pkharchenko/R/x86_64-pc-linux-gnu-library/3.1",.libPaths()))
library(scde); library(WGCNA)
```
### Preparing data
The read count table and cell group annotations from _Pollen et al._ can be loaded using {{data("pollen") call}}. Some additional filter sare applied:
```{r data,cache=TRUE,hide=TRUE,eval=TRUE}
  # read in the expression matrix
  load("pollen.rda")

  cd <- pollen.counts
    
  # filter data
  # filter out low-gene cells
  vi <- colSums(cd>0)>1.8e3; table(vi)
  #hist(colSums(cd>0))
  cd <- cd[,vi]; 
  
  # remove genes that don't have many reads
  vi <- rowSums(cd)>10; table(vi)
  cd <- cd[vi,];

  # remove genes that are not seen in a sufficient number of cells
  vi <- rowSums(cd>0)>5; table(vi)
  cd <- cd[vi,];

  # check the final dimensions of the read count matrix
  dim(cd)
```

Next, we'll translate group and sample source data from Pollen et al into color codes:
```{r colorcodes}
l1cols <- c("red","orange","blue","green")[as.integer(as.factor(pollen.groups[colnames(cd)]))]
# sample sources
x <- gsub("^Hi_(.*)_.*","\\1",colnames(cd))
l2cols <- c("coral4","olivedrab3","skyblue2","slateblue3")[as.integer(factor(x,levels=c("NPC","GW16","GW21","GW21+3")))]

```

```{r dataload,include=FALSE}
#load("cd.RData")
```

### Fitting error models
Next, we'll construct error models for individual cells. Here we'll use k-nearest neighbour model fitting procedure implemented by knn.error.models() method. This is a relatively noisy dataset (non-UMI), so we raise the min.count.threshold to 2 (minimum number of reads for the gene to be initially classified as a non-failed measurement), requiring at least 5 non-failed measurements per gene. We're providing a rough guess to the complexity of the population, by fitting the error models based on 1/4 of most similar cells (i.e. gussing there might be ~4 subpopulations). 

```{r models,cache=TRUE,hide=TRUE,warning=FALSE,message=FALSE}
n.cores <- 10; # number of parallel cores to use
knn <- knn.error.models(cd,k=ncol(cd)/4,n.cores=n.cores,min.count.threshold=2,min.nonfailed=5,max.model.plots=10)
```

The fitting process above wrote out cell.models.pdf file in the current directory showing model fits for the first 10 cells (see max.model.plots argument). Here's an example of such plot:
![cell 3 model](cell.model.fits-0.png)

The two scatter plots on the left show observed (in a given cell) vs. expected (from k similar cells) expression magnitudes for each gene that is being used for model fitting. The second (from the left) scatter plot shows genes belonging to the drop-out component in red. The black dashed lines show 95% confidence band for the amplified genes (the grey dashed lines show confidence band for an alternative constant-theta model). The third plot shows drop-out probability as a function of magnitude, and the fourth plot shows negative binomial theta local regression fit as a function of magnitude (for the amplified component). 

### Normalizing variance
Here we normalize variance, trimming 3 most extreme cells and limiting maximum adjusted variance to 5.
```{r varnorm,fig.height=3,fig.width=6,cache=TRUE}
varinfo <- pagoda.varnorm(knn,counts=cd,trim=3/ncol(cd),max.adj.var=5,n.cores=n.cores,plot=T)
```

The plot on the left shows coefficient of variance squared (on log10 scale) as a function of expression magntidue (log10 FPM). The red line shows local regression model for the genome-wide average dependency. The plot on the right shows adjusted variance (derived based on chi-squared probability of observed/genomewide expected ratio for each gene, with degrees of freedom adjusted for each gene). The adjusted variance of 1 means that a given gene exchibits as much variance as expected for a gene of such population average expression magnitude. Genes with high adjusted variance are overdispersed within the measured population and most likely show subpopulation-specific expression:

```{r}
# list top overdispersed genes
sort(varinfo$arv,decreasing=T)[1:10]
```

### Controling for sequencing depth differences
Even with all the corrections, sequencing depth or gene coverage is typically still a major aspects of variability. In most studies, we would want to control for that as a technical artifact (exceptions are cell mixtures where subtypes significantly differ in the amount of total mRNA). Below we will control for the gene coverage (estimated as a number of genes with non-zero magnitude per cell) and normalize out that aspect of cell heterogeneity: 
```{r cache=TRUE}
varinfo <- pagoda.subtract.aspect(varinfo,colSums(cd[,rownames(knn)]>0))
```


### Evaluate overdispersion of pre-defined gene sets
For pre-defined gene sets we'll use GO annotations. To do so we'll construct an environment mapping GO terms to the set of genes contained in it.
```{r goenv,cache=TRUE}
library(org.Hs.eg.db)
# translate gene names to ids
ids <- unlist(lapply(mget(rownames(cd),org.Hs.egALIAS2EG,ifnotfound=NA),function(x) x[1]))
# reverse map
rids <- names(ids); names(rids) <- ids;
# list all the ids per GO category
go.env <- eapply(org.Hs.egGO2ALLEGS,function(x) as.character(na.omit(rids[x])))
 # omit categories with too few genes
go.env <- go.env[unlist(lapply(go.env,length))>5];
  
# append descriptions to the GO names
library(GO.db)
desc <- unlist(lapply(mget(names(go.env),GOTERM,ifnotfound=NA),function(x) if(is.logical(x)) { return("") } else { slot(x,"Term")}))
names(go.env) <- paste(names(go.env),desc); # append description to the names
go.env <- list2env(go.env); # convert to an environment
```

Calculate weighted PC magnitudes of GO gene sets:
```{r pathwaySig,cache=TRUE}
pwpca <- pagoda.pathway.wPCA(varinfo,go.env,n.components=1,n.cores=n.cores,n.internal.shuffles=50)
```

We can now evaluate the statistical significance of the observed overdispersion for each gene set. 
```{r topPathways,cache=TRUE,fig.height=4,fig.width=5}
df <- pagoda.top.aspects(pwpca,return.table=T,plot=T,z.score=1.96)
```

Each point on the plot shows the PC1 variance (lambda1) magtnidue (normalized by set size) as a function of set size. The red lines show expected (solid) and 95% upper bound (dashed) magnitudes based on the Tracey-Widom model.

```{r}
head(df)
```

* The z column gives the Z-score of pathway over-dispersion realtive to the genome-wide model (Z-score of 1.96 corresponds to P-value of 5%, etc.). 
* "z.adj" column shows the Z-score adjusted for multiple hypothesis (using Benjamini-Hochberg correction). 
* "score" gives observed/expected variance ratio
* "sh.z" and "adj.sh.z" columns give the raw and adjusted Z-scores of "pathway cohesion", which compares the observed PC1 magnitude to the magnitudes obtained when the observations for each gene are randomized with respect to cells. When such Z-score is high (e.g. for GO:0008009) then multiple genes within the pathway contribute to the coordinated pattern.


### Determine gene clusters and their overdispersion significance
The following procedure will determine gene clusters in the data, and biuld a background model for the expectation of the gene cluster weighted PC magnitudes. Note higher trim values for the clusters, as we want to avoid clusters that are formed by outlier cells.
```{r clusterPCA,cache=TRUE,fig.height=3,fig.width=6,warning=FALSE,message=FALSE}
clpca <- pagoda.gene.clusters(varinfo,trim=7.1/ncol(varinfo$mat),n.clusters=150,n.cores=n.cores,plot=T)
```

The plot above shows background distribution of the PC1 variance (lambda1) magnitude. The blue scatterplot on the left shows lambda1 magnitude vs. cluster size for clusters determined based on randomly-generated matrices of the same size. The black circles show top cluster in each simulation. The red lines show expected magnitude and 95% confidence interval based on Tracy-Widom distribution. The right plot shows extreme value distribution fit of residual cluster PC1 variance magnitude relative to the Tracy-Widom distribution.

Now the set of top aspects can be recalculated taking gene clusters into account:
```{r topPathways2,cache=TRUE,fig.height=4,fig.width=5}
df <- pagoda.top.aspects(pwpca,clpca,return.table=T,plot=T,z.score=1.96)
head(df)
```
The gene clusters and their corresponding model expected value and 95% upper bound are shown in green.


### Visualize significant aspects of heterogeneity
To view top heterogeneity aspects, we will first obtain a full info on the significant aspects, and determine overall cell clustering:
```{r celclust,cache=TRUE}
# get full info on the top aspects
tam <- pagoda.top.aspects(pwpca,clpca,,z.score=qnorm(0.01/2,lower.tail=F))
# determine overall cell clustering
hc <- pagoda.cluster.cells(tam,varinfo)
```

Next, we will reduce redudant aspects in two steps. First we will combine pathways that are driven by the same sets of genes:
```{r loadingCollapse,cache=T}
tamr <- pagoda.reduce.loading.redundancy(tam,pwpca,clpca)
```

In the second step we will combine aspects that show similar patterns (i.e. separate the same sets of cells). Here we will plot the cells using the overall cell clustering determined above:
```{r correlatedCollapse,cache=T,fig.height=6,fig.width=10}
tamr2 <- pagoda.reduce.redundancy(tamr,distance.threshold=0.9,plot=T,cell.clustering=hc,labRow=NA,labCol=NA,box=T,margins=c(0.5,0.5),trim=0)
```

In the plot above, the columns are cells, rows are different significant aspects, clustered by their similarity pattern.The green-to-orange color scheme shows low-to-high weighted PCA scores (aspect patterns), where generally orange indicates higher expression. Blocks of color on the left margin show which aspects have been combined by the command above. Here the number of resulting aspects is relatively small. "top" argument (i.e. top=10) can be used to limit further analysis to top N aspects.

We will view top aspects, clustering them by pattern similarity (note, to view aspects in the order of increasing lambda1 magnitude, use row.clustering=NA). 
```{r viewAspects,fig.height=3.5,fig.width=8}
col.cols <- rbind(groups=cutree(hc,3))
pagoda.view.aspects(tamr2,cell.clustering=hc,box=T,labCol=NA,margins=c(0.5,15),col.cols=rbind(l1cols))
```

While each row here represents a cluster of pathways, the row names are assigned to be the top overdispersed aspect in each cluster.

To pull up an interactive browser interface, we will create a PAGODA app:
```{r pagodaApp,eval=FALSE}
# compile a browsable app, showing top three clusters with the top color bar
app <- make.pagoda.app(tamr2,tam,varinfo,go.env,pwpca,clpca,col.cols=col.cols,cell.clustering=hc,title="NPCs")
  
# show app in the browser (port 1468)
show.app(app,"pollen",browse=T,port=1468)  
saveRDS(app,file="pollen2.app.rds")
```

The browser app allows to view the gene sets grouped within each aspect (row), as well as genes underlying the detected heterogeneity patterns. Similar views can be obtained in the R session itself. For instance, here we'll view top 10 genes associated with the top two pathways in the neurogenesis cluster: "neurogenesis" (GO:0022008) and "generation of neurons" (GO:0048699)
```{r showTopPathwayGenes,cache=TRUE,fig.height=3.5,fig.width=8}
pagoda.show.pathways(c("GO:0022008 neurogenesis","GO:0048699 generation of neurons"),varinfo,go.env,cell.clustering=hc,margins=c(1,5),show.cell.dendrogram=T,showRowLabels=T,showPC=T)
```

### Controlling for undesired aspects of heterogeneity
To explicitly control for aspects of heterogeneity that are not of interest for this particular experiment we will use pagoda.subtract.aspect method that we've previously used to control for residual patterns associated with sequencing depth differences. Here we illustrate how to control for the mitotic cell cycle pattern which showed up as one of the four significant aspects in the analysis above.

```{r controlForCellCycle,cache=TRUE}
# get cell cycle signature and view the top genes
cc.pattern <- pagoda.show.pathways(c("GO:0000280 nuclear division","GO:0007067 mitotic nuclear division"),varinfo,go.env,show.cell.dendrogram=T,cell.clustering=hc,showRowLabels=T)
# subtract the pattern
varinfo.cc <- pagoda.subtract.aspect(varinfo,cc.pattern)
```
Now we can go through the same analysis as shown above, starting with the pagoda.pathway.wPCA() call, using varinfo.cc instead of varinfo, which will control for the cell cylce heterogeneity between the cells.
